### PR TITLE
Bug 1381623 - A new Photon menu. Behind a feature flag. Disabled by default

### PR DESCRIPTION
**Pull request by farhanpatel**
This is a new menu to replace the current grid menu we use in the app. The UI is nowhere near final but functionally this is how the menu is going to be. I want to land the menu early and then iterate on the UI. 
screen shot 2017-07-18 at 1 36 44 pm

You can enable the menu by going to the debug menu and turning on "Photon Menu". This is nowhere near done so it is also disabled for nightly/debug. Everyone should continue to see the current menu design. 

### About the structure of the code. 
- I opted to create a PhotonActionSheetProtocol which defines all the functionality of the menu. Both BVC and TabTrayController should be able to conform to this protocol. This removes the giant `performMenuAction` from BVC and TabTray.  I want to remove as much code out of the BVC as possible. 
- I decided to not use AppState and once we remove the current menu I'll remove appState. The state of the menu is computed when it is opened and it cannot change once it has opened. 

ps: I had to fix a bug in FeatureSwitches in order to make sure that features that are disabled to all audiences can still be enabled manually. Allowing us to land features like this behind a flag.

### CODE REVIEW

**Comment by farhanpatel** [comment]
### SwiftLint found issues

#### Warnings

File | Line | Reason |
| --- | ----- | ----- |
BrowserViewController.swift | 1808 | Limit vertical whitespace to a single empty line. Currently 2. 
BrowserViewController.swift | 1811 | Limit vertical whitespace to a single empty line. Currently 2. 
BrowserViewController.swift | 1815 | Limit vertical whitespace to a single empty line. Currently 2. 
BrowserViewController.swift | 1474 | Arguments can be omitted when matching enums with associated types if they are not used. 
BrowserViewController.swift | 1476 | Arguments can be omitted when matching enums with associated types if they are not used. 
BrowserViewController.swift | 1795 | Prefer `-> Void` over `-> ()`. 
AppSettingsOptions.swift | 74 | Colons should be next to the identifier when specifying a type and next to the key in dictionary literals. 
AppSettingsOptions.swift | 587 | Colons should be next to the identifier when specifying a type and next to the key in dictionary literals. 
AppSettingsOptions.swift | 187 | Arguments can be omitted when matching enums with associated types if they are not used. 
PhotonActionSheet.swift | 343 | Limit vertical whitespace to a single empty line. Currently 2. 
PhotonActionSheetProtocol.swift | 96 | Limit vertical whitespace to a single empty line. Currently 2. 
PhotonActionSheetProtocol.swift | 146 | Limit vertical whitespace to a single empty line. Currently 2. 
PhotonActionSheetProtocol.swift | 1 | Files should have consistent header comments. 
PhotonActionSheetProtocol.swift | 35 | Prefer `-> Void` over `-> ()`. 
PhotonActionSheetProtocol.swift | 97 | Prefer `-> Void` over `-> ()`. 
PhotonActionSheetProtocol.swift | 151 | Prefer `-> Void` over `-> ()`. 
#### Errors

File | Line | Reason |
| --- | ----- | ----- |
PhotonActionSheetProtocol.swift | 97 | There should be no space before and one after any comma. 
PhotonActionSheetProtocol.swift | 9 | Opening braces should be preceded by a single space and on the same line as the declaration. 


<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

**Comment by farhanpatel** [comment]
Darn! Turns out I'm using a feature of swift 4 (which I have turned on in xcode9) 
Sorry about the breaking build I'll fix that.

File: Client.xcodeproj/xcshareddata/xcschemes/Fennec.xcscheme [hunk 1]
```diff
@@ -26,6 +26,7 @@
       buildConfiguration = "Fennec"
       selectedDebuggerIdentifier = "Xcode.DebuggerFoundation.Debugger.LLDB"
       selectedLauncherIdentifier = "Xcode.DebuggerFoundation.Launcher.LLDB"
+      language = ""
```
Comment (garvankeeley) [inline Client.xcodeproj/xcshareddata/xcschemes/Fennec.xcscheme]: is this change needed?

**Comment by garvankeeley** [comment]
A change to pull some extensions out of BVC just landed on master which will cause merge conflicts, sorry about the hassle
